segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,378 @@
+            int bytesRead = channel.read(ByteBuffer.wrap(buffer, bufferOffset, length), offset);
+            return bytesRead;
+        } catch (IOException | IndexOutOfBoundsException e) {
+            if (e instanceof IndexOutOfBoundsException) {
+                throw new CompletionException(new ArrayIndexOutOfBoundsException(e.getMessage()));
+            } else {
+                throw new CompletionException(e);
+            }
+        }
+    }
+
+    private SegmentProperties syncGetStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        try {
+            PosixFileAttributes attrs = Files.readAttributes(Paths.get(config.getNfsRoot(), streamSegmentName),
+                    PosixFileAttributes.class);
+            StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName, attrs.size(),
+                    !(attrs.permissions().contains(OWNER_WRITE)), false,
+                    new ImmutableDate(attrs.creationTime().toMillis()));
+
+            return information;
+        } catch (IOException ioe) {
+            throw new CompletionException(ioe);
+        }
+    }
+
+    private boolean syncExists(String streamSegmentName, Duration timeout) {
+        boolean exists = Files.exists(Paths.get(config.getNfsRoot(), streamSegmentName));
+        return exists;
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName, Duration timeout) {
+        log.info("Creating Segment {}", streamSegmentName);
+        try {
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as rw-r--r-- 644
+            perms.add(PosixFilePermission.OWNER_WRITE);
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            Files.createDirectories(path.getParent());
+            Files.createFile(path, fileAttributes);
+            log.info("Created Segment {}", streamSegmentName);
+            return this.getStreamSegmentInfo(streamSegmentName, timeout).get();
+        } catch (IOException | InterruptedException | ExecutionException e) {
+            log.info("Exception {} while creating a segment {}", e, streamSegmentName);
+            if (e instanceof FileAlreadyExistsException) {
+                throw new CompletionException(new StreamSegmentExistsException(streamSegmentName, e));
+            } else {
+                throw new CompletionException(e);
+            }
+        }
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length, Duration timeout) {
+        log.trace("Writing {} to segment {} at offset {}", length, handle.getSegmentName(), offset);
```

Comment:
Use standard `LoggerHelpers.traceEnter/traceEnterWithContext + traceLeave` instead of this ad-hoc messaging. Here and with every other method in this class (even those that are missing it).